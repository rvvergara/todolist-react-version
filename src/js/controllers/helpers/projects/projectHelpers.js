import uuid from 'uuid';

import projectsController from '../../projectsController';

import * as localStorageData from '../common/storage';

export const createProjectsArray = () => {
  if (localStorageData.getDataFromLocalStorage('projectsArray') === null) {
    const projectsArray = [];
    localStorageData.setDataIntoLocalStorage('projectsArray', projectsArray);
  }
};

// Create a default project for new users or after localstorage is cleared
export const createDefaultProject = () => {
  if (localStorageData.getDataFromLocalStorage('projectCount') === 0) {
    const defaultProject = projectsController.create('Default Project', uuid());
  }
  if (localStorageData.getDataFromLocalStorage('Default Project') !== null) {
    const defaultProject = localStorageData.getDataFromLocalStorage('Default Project');
  }
};

// Adding event listener to Project Form
// document.getElementById('projectNameForm').addEventListener('submit', e => {
//   e.preventDefault();
//   submitProjectForm(e.target);
// });
// // Adding event listener to Add Project Btn
// document.getElementById('addProjBtn').addEventListener('click', e => {
//   const action = 'new';
//   showProjectForm(e.target, action);
// });
