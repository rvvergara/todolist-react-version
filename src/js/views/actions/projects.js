import uuid from 'uuid';

const getProjects = projects => ({
  type: 'GET_PROJECTS',
  projects,
});

const addProject = ({ name = 'PROJECT', description = 'SOME PROJECT' } = {}) => ({
  type: 'ADD_PROJECT',
  project: {
    id: uuid(),
    name,
    description,
    todos: [],
  },
});

const updateProject = (id, updates) => ({
  type: 'UPDATE_PROJECT',
  id,
  updates,
});

const deleteProject = id => ({
  type: 'DELETE_PROJECT',
  id,
});

export {
  getProjects,
  addProject,
  updateProject,
  deleteProject,
};
