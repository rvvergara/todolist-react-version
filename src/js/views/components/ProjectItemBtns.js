import React from 'react';
import PropTypes from 'prop-types';

const ProjectItemBtns = (props) => {
  const { project, handleDeleteProject, clickUpdateProjectBtn } = props;
  return (
    <div>
      <button
        type="button"
        className="btn btn-sm btn-danger"
        id={`delete-${project.id}`}
        onClick={handleDeleteProject}
      >
        Delete
      </button>
      <button
        type="button"
        id={`update-${project.id}`}
        className="btn btn-sm btn-warning"
        onClick={clickUpdateProjectBtn}
      >
        Update
      </button>
    </div>
  );
};

ProjectItemBtns.propTypes = {
  project: PropTypes.instanceOf(Object).isRequired,
  handleDeleteProject: PropTypes.func.isRequired,
  clickUpdateProjectBtn: PropTypes.func.isRequired,
};

export default ProjectItemBtns;
