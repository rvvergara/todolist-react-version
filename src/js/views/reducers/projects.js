export default (state = [], action) => {
  switch (action.type) {
    case 'GET_PROJECTS':
      return [...state, ...action.projects]

    case 'ADD_PROJECT':
    {
      const index = state.findIndex(project => project.name === action.project.name);
      return index === -1 ? state.concat(action.project) : state;
    }

    case 'UPDATE_PROJECT':
    {
      const { name } = action.updates;
      const conflictIndex = state.findIndex(project => project.name === name);
      const projectIndex = state.findIndex(project => project.id === action.id);
      const updatedProject = { ...state[projectIndex], ...action.updates };
      return conflictIndex === -1 ? [...state.slice(0, projectIndex), updatedProject, ...state.slice(projectIndex + 1)] : state;
    }
    case 'DELETE_PROJECT':
      return state.filter(project => project.id !== action.id);
    default:
      return state;
  }
};
