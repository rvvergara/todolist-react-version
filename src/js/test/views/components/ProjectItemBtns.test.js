import React from 'react';
import { shallow } from 'enzyme';
import ProjectItemBtns from '../../../views/components/ProjectItemBtns';
import projects from '../../fixtures/projects';

let wrapper;
let project;
let handleDeleteProject;
let clickUpdateProjectBtn;

beforeEach(() => {
  project = projects[0];
  handleDeleteProject = jest.fn();
  clickUpdateProjectBtn = jest.fn();
  wrapper = shallow(
    <ProjectItemBtns
      project={project}
      handleDeleteProject={handleDeleteProject}
      clickUpdateProjectBtn={clickUpdateProjectBtn}
    />,
  );
});

test('should render correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('clickUpdateProjectBtn should be called on click to Update btn', () => {
  wrapper.find('button').at(1).simulate('click');
  expect(clickUpdateProjectBtn).toHaveBeenCalled();
});

test('handleDelete should be called on click to Delete btn', () => {
  wrapper.find('button').at(0).simulate('click');
  expect(handleDeleteProject).toHaveBeenCalled();
});
