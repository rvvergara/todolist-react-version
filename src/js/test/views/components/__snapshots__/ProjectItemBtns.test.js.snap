// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  <button
    className="btn btn-sm btn-danger"
    id="delete-0"
    onClick={[MockFunction]}
    type="button"
  >
    Delete
  </button>
  <button
    className="btn btn-sm btn-warning"
    id="update-0"
    onClick={[MockFunction]}
    type="button"
  >
    Update
  </button>
</div>
`;
